File: graph/Bellman-Ford.cpp
#include <bits/stdc++.h>
using namespace std;

const int MAX_E=1000;
const int MAX_V=1000;

#define INF (1 << 29)

//
struct edge { int from, to, cost; };

edge es[MAX_E]; // 辺

int d[MAX_V];
int V, E;

//
void bellman_ford(int s){
    //
    for(int i=0; i<V; i++) d[i]=INF;
    //
    d[s] = 0;
    while(true){
        bool update=false;
        //
        for(int i=0; i<E; i++){
            //
            edge e = es[i];
            cout << d[e.from] <<" "<< d[e.to] <<" "<< e.cost <<"  ";
            //
            if(d[e.from] != INF && d[e.to] > d[e.from] + e.cost){
                //
                d[e.to] = d[e.from] + e.cost;
                //
                update = true;
            }
        }
        cout << endl;
        //
        if(!update) break;
    }
    
    cout << d[E-1] << endl;
}

//
//
bool find_negative_loop(){
    //
    memset(d, 0, sizeof(d));
    
    //
    for(int i=0; i<V; i++){
        for(int j=0; j<E; j++){
            edge e = es[j];
            if(d[e.to]>d[e.from]+e.cost){
                d[e.to]=d[e.from]+e.cost;
                
                // 
                if(i==V-1) return true;
            }
        }
    }
    return false;
}

int main(){
    cin >> V >> E;
    int from, to, cost;
    for(int i=0; i<E; i++){
        cin >> from >> to >> cost;
        es[i] = (edge){from,to,cost};
    }
    bellman_ford(0);
}

File: library/rust/src/graph/bellman_ford.rs
// unverified

use std;

const MAX_E:usize = 2010;
const MAX_V:usize = 1010;
const INF:i64 = std::i64::MAX;

#[derive(Debug, Clone, Copy)]
struct Edge {
    from: usize,
    to: usize,
    cost: i64
}

// 単一始点最短経路 (負のコストを含む). 負の閉路が存在したら true を返す
fn bellman_ford(s:usize, edges:&Vec<Edge>, d:&mut Vec<i64>, V:usize, E:usize) -> bool {
    let mut cnt = 0;
    // 最短経路の初期化
    let mut d = vec![INF;V];
    d[s] = 0;
    loop {
        cnt += 1;
        let mut update = false;
        // 全ての辺に対して
        for i in 0..E {
            let e = edges[i];
            // この時点での、e.toへ行くコストより、e.fromを経由して行く方がコストが小さい時、d[e.to]を更新する
            if d[e.from] != INF && d[e.to] > d[e.from] + e.cost {
                d[e.to] = d[e.from] + e.cost;
                // V回目のループで更新が行われた場合、負の閉路が存在する
                if cnt == V as i64 { return true; }
                update = true;
            }
        }
        // 更新がなければ終了
        if !update { break; }
    }
    false
}